### PR TITLE
Remove KOKKOS_IMPL_CUDA_* macros

### PR DESCRIPTION
**Pull request by masterleinad**
Remove most KOKKOS_IMPL_CUDA_* macros. These are internal (and short) and only really have one definition anyway.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

File: .jenkins [hunk 1]
```diff
@@ -36,7 +36,7 @@ pipeline {
                         sh 'ccache --zero-stats'
                         sh '''rm -rf build && mkdir -p build && cd build && \
                               cmake \
-                                -DCMAKE_BUILD_TYPE=Release \
+                                -DCMAKE_BUILD_TYPE=Debug \
```
Comment (crtrott) [inline .jenkins:39]: ??

File: core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp [hunk 2]
```diff
@@ -388,48 +368,28 @@ __device__ inline
...
       if (int(blockDim.x * blockDim.y) > 2) {
         value_type tmp = Kokkos::shfl_down(value, 2, 32);
         if (id + 2 < int(gridDim.x)) reducer.join(value, tmp);
       }
-#ifdef KOKKOS_IMPL_CUDA_SYNCWARP_NEEDS_MASK
-      active += KOKKOS_IMPL_CUDA_BALLOT_MASK(mask, 1);
-#else
-      active += KOKKOS_IMPL_CUDA_BALLOT(1);
-#endif
+      active += __ballot_sync(__activemask(), 1);
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp:377]: this is wrong, this must use the previously established mask.

**Review by crtrott — CHANGES_REQUESTED** [review]
There is one mistake. Otherwise looks ok. But I want another *careful* review. Screwing this up, like that one mistake, would mostly manifest as super rare race conditions....

File: core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp]: Fixed in [`3d32df0` (#4138)](https://github.com/kokkos/kokkos/pull/4138/commits/3d32df071803f492a4c0c8fc460ec92f0d7ca6e0)

File: .jenkins (hunk 1 above)
Comment (masterleinad) [reply .jenkins]: Fixed in [`3d32df0` (#4138)](https://github.com/kokkos/kokkos/pull/4138/commits/3d32df071803f492a4c0c8fc460ec92f0d7ca6e0).

**Comment by masterleinad** [comment]
Retest this please.

**Review by crtrott — APPROVED** [review]

**Comment by dalg24** [comment]
Please elaborate what changes are proposed and why you suggested them.  (I don't have an issue with them, just with the way you present them)

**Comment by masterleinad** [comment]
> Please elaborate what changes are proposed and why you suggested them. (I don't have an issue with them, just with the way you present them)

Since requiring CUDA version >9.0, these macros only have one definition and there is no expectation that we need to provide alternatives for other versions. Thus, this pull request is solely about code cleanup removing branches that will never be executed, and avoiding macros that hide the actual code being used. The specific macros being considered are:
- KOKKOS_IMPL_CUDA_ACTIVEMASK
- KOKKOS_IMPL_CUDA_SYNCWARP
- KOKKOS_IMPL_CUDA_SYNCWARP_MASK
- KOKKOS_IMPL_CUDA_BALLOT
- KOKKOS_IMPL_CUDA_BALLOT_MASK
- KOKKOS_IMPL_CUDA_SHFL
- KOKKOS_IMPL_CUDA_SHFL_MASK
- KOKKOS_IMPL_CUDA_SHFL_UP
- KOKKOS_IMPL_CUDA_SHFL_DOWN
- KOKKOS_IMPL_CUDA_SHFL_DOWN_MASK

The other macros in `core/src/Cuda/Kokkos_Cuda_Version_9_8_Compatibility.hpp` seem useful. The name of the file might be a misnomer now, though.

Additionally, some tests in `core/unit_test/TestViewMapping_subview.hpp` are reenabled for CUDA >10.0. The intent before was to disable them for CUDA 9 without taking newer versions into account.

**Review by nliber — APPROVED** [review]

**Comment by dalg24** [comment]
> The other macros in `core/src/Cuda/Kokkos_Cuda_Version_9_8_Compatibility.hpp` seem useful.

Actually we can probably get rid of everything.

`KOKKOS_IMPL_CUDA_MAX_SHFL_SIZEOF` is not actually used anywhere.
https://github.com/kokkos/kokkos/blob/1fb0c284d458c75370094921d9f202c287502325/core/src/Cuda/Kokkos_Cuda_Vectorization.hpp#L138-L141

`KOKKOS_IMPL_CUDA_SYNCWARP_OR_RETURN` is exclusively used in `core/src/Cuda/Kokkos_Cuda_Task.hpp` and all occurrences are actually commented except https://github.com/kokkos/kokkos/blob/1fb0c284d458c75370094921d9f202c287502325/core/src/Cuda/Kokkos_Cuda_Task.hpp#L393-L395.
At the very least we could move the macro definition closer to its point of use.

**Comment by ndellingwood** [comment]
@masterleinad @dalg24 The `KOKKOS_IMPL_CUDA_BALLOT` macro was used in Trilinos by Sacado, merge of this PR caused nightly Trilinos builds to fail. Is the correct work around to see about adding macro to Sacado?

```
#if defined(__CUDA_ARCH__)
#define KOKKOS_IMPL_CUDA_BALLOT(x) __ballot_sync(__activemask(), x)
#else
define KOKKOS_IMPL_CUDA_BALLOT(x) 0
#endif
```

**Comment by masterleinad** [comment]
Yes, `KOKKOS_IMPL_CUDA_BALLOT(x)` should be replaced by `__ballot_sync(__activemask(), x)`.

File: core/unit_test/TestViewMapping_subview.hpp [hunk 3]
```diff
@@ -81,7 +81,7 @@ struct TestViewMappingSubview {
   using DLT  = Kokkos::View<int** * [13][14], Kokkos::LayoutLeft, ExecSpace>;
   using DLS1 = Kokkos::Subview<DLT, range, int, int, int, int>;
 
-#if !defined(KOKKOS_IMPL_CUDA_VERSION_9_WORKAROUND)
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 1000
```
Comment (dalg24) [inline core/unit_test/TestViewMapping_subview.hpp:84]: ???

**Pull request merged**